pkg/kubelet/network/hostport/fake_iptables.go

```diff
@@ -274,6 +275,11 @@ func (f *fakeIPTables) restore(restoreTableName utiliptables.Table, data []byte,
 					}
 				}
 				_, _ = f.ensureChain(tableName, chainName)
+				if !strings.Contains(allLines, "-X "+string(chainName)) {
+					if err := f.FlushChain(tableName, chainName); err != nil {
```

Comment:
Sorry I don't quite get the logic here, why should we flush the chain when there is no `-X CHAIN_NAME`? Explanation would be appreciated.

Reply:
I think I should update here to just flush user defined chains. 
The --noflush option for iptables-restore doesn't work for user-defined chains, such as TESTCHAIN, only builtin chains. https://unix.stackexchange.com/questions/134687/how-to-combine-iptables-rulesets

I did a test to confirm this. See below example, the POSTROUTING chain didn't get flush, but KUBE-NODEPORT chain did.
```
[root@kubernetes-master vagrant]# iptables -t nat -S
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
-N KUBE-HP-5N7UH5JAXCVP5UJR
-N KUBE-NODEPORT
-A POSTROUTING -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
-A KUBE-NODEPORT -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
[root@kubernetes-master vagrant]# cat nat.log 
# Generated by iptables-save v1.4.21 on Mon Nov 13 03:14:59 2017
*nat
:PREROUTING ACCEPT [65:3900]
:INPUT ACCEPT [65:3900]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:KUBE-HP-5N7UH5JAXCVP5UJR - [0:0]
:KUBE-NODEPORT - [0:0]
COMMIT
# Completed on Mon Nov 13 03:14:59 2017
[root@kubernetes-master vagrant]# iptables-restore --table=nat --noflush < nat.log 
[root@kubernetes-master vagrant]# iptables -t nat -S
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
-N KUBE-HP-5N7UH5JAXCVP5UJR
-N KUBE-NODEPORT
-A POSTROUTING -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
```

And this behavior is needed to cleanup rules in KUBE-NODEPORT in my tests.

Reply:
Thanks for the investigation, now I get it. Interesting that this behavior is not documented in iptables manual.

Though I just learned that user-defined chains won't always be flushed, unless they are explicitly mentioned in the input of `iptables-restore --noflush`, as what below comment is referring to:
https://github.com/kubernetes/kubernetes/blob/f28140429e0f3b6b4ef1bb7399d1bbd7d644c933/pkg/kubelet/network/hostport/hostport_syncer.go#L263-L267

In short I believe your fix will work as expected, but the implementation of `fakeIPTables.restore()` still seems problematic as it flushes more than it should...

Reply:
Oops I scan too fast, looks like that is exactly what you have implemented. Could you update it to just flush user defined chains? Thanks.

Reply:
Done. Also add a test `TestRestoreFlushRules` for `fackIPtables.restore()`.

Reply:
Awesome, thanks!